src/scene/materials/shader-graph-node.js

```diff
@@ -0,0 +1,615 @@
+                            }
+                        }
+                    }
+
+                    depGraphFuncs[name] = subGraph._generateFuncGlsl();
+                    subGraph._generateSubGraphFuncs(depGraphFuncs, depIoPortList);
+                }
+            }
+        }
+    },
+
+    generateRootDeclGlsl: function () {
+        var i;
+        var ioPort;
+        var generatedGlsl = '';
+        // run through inputs (and const sampler2Ds) to declare uniforms - (default) values are set elsewhere
+        for (i = 0; i < this.graphData.ioPorts.length; i++) {
+            var matId = '_' + this.id;
+            ioPort = this.graphData.ioPorts[i];
+            if (ioPort.name.startsWith('IN_') || (ioPort.name.startsWith('CONST_') && ioPort.type === 'sampler2D')) {
+                generatedGlsl += 'uniform ' + ioPort.type + ' ' + ioPort.name + matId + ';\n';
+            }
+        }
+        // run through constants values are set here (except for textures - which have to be uniforms)
+        for (i = 0; i < this.graphData.ioPorts.length; i++) {
+            ioPort = this.graphData.ioPorts[i];
+            if (ioPort.name.startsWith('CONST_') && (ioPort.type != 'sampler2D' )) {
+                generatedGlsl += ioPort.type + ' ' + ioPort.name + ' = ' + this._getIoPortValueString(ioPort) + ';\n';
+            }
+        }
+
+        // get all sub graph function definitions (including functions in sub graphs' sub graphs ...)
+        // assumes names are unique - maybe should be id or key?
+        var depGraphFuncs = {};
+        var depIoPortList = [];
+
+        var depName = this.name;
+        if (!this.graphData.customFuncGlsl) {
+            depName += '_' + this.id;
+        }
+        depGraphFuncs[depName] = this._generateFuncGlsl(); // this should prevent infinite recursion?
+
+        this._generateSubGraphFuncs(depGraphFuncs, depIoPortList);
+
+        // declare all dependancy textures
+        for (i = 0; i < depIoPortList.length; i++) {
+            generatedGlsl += depIoPortList[i];// .declString;
+        }
+
+        // add all the graph definitions
+        var depGraphList = []; // reverse order
+        for (var func in depGraphFuncs) {
+            var funcString = '';
+
+            if ( func.endsWith('PS') ) {
+                funcString += '#ifdef SG_PS\n';
+            } else if ( func.endsWith('VS') ) {
+                funcString += '#ifdef SG_VS\n';
+            }
+
+            funcString += depGraphFuncs[func] + '\n';
+
+            if ( func.endsWith('PS')  ) {
+                funcString += '#endif //SG_PS\n';
+            } else if ( func.endsWith('VS') ) {
+                funcString += '#endif //SG_VS\n';
+            }
+
+            depGraphList.push(funcString);
+        }
+        var tLen = depGraphList.length; // need this because pop() reduces array length!
+        for (i = 0; i < tLen; i++) {
+            generatedGlsl += depGraphList.pop();
+        }

```

Comment:
while (depGraphList.length) {
    generatedGlsl += depGraphList.pop();
}

Reply:
nice - I'll swap in this code - thanks!

Reply:
done - thanks!